### PR TITLE
Fix element memoization checks. Allows placeholder to change dynamically based on the props. Allows renderLeaf to be memoized.

### PR DESCRIPTION
**Pull request by ryanmitts**
There are bugs with the memoization checks in the element, text, and leaf components.

The `renderLeaf` function to the Editable cannot be memoized as if it is, the rendering of leaves are not reconciled correctly on the DOM as the old node is allowed to stay despite the leaf being different. This is because the leaf element does not actually check it's text content in the memoization check.

The placeholder decorator is not re-rendered when the placeholder text is changed, this is because the element, leaf and text components do not check the decorator properties.

#### Is this adding or improving a _feature_ or fixing a _bug_?

Bug.

#### What's the new behavior?

The appropriate components check in the memoization comparison if two decorator range list items are placeholders, and if they are, if their text equals.

Demo

You will also be able to memoize the `renderLeaf` function when using decorators.

#### How does this change work?

Checking the decorator props, and the leaf checking it's text is the same in the memoization checks.

#### Have you checked that...?

- [x] The new code matches the existing patterns and styles.
- [x] The tests pass with `yarn test`.
- [x] The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
- [x] The relevant examples still work. (Run examples with `yarn start`.)

#### Does this fix any issues or need any specific reviewers?

Fixes: #3423, #3432, #3447
Reviewers: @

### CODE REVIEW

**Comment by Morphexe** [comment]
Shouldnt this be merged ? This is actually a big issue currently.

**Comment by kamilmielnik** [comment]
Bump for justice.

**Comment by bithavoc** [comment]
This PR is quite a big deal, would love to have this merged.

**Comment by kamilmielnik** [comment]
Bumping for justice.

We've had to implement a custom placeholder component because of this issue.
I really would appreciate it if it got into the next release.

**Comment by haderman** [comment]
would be really nice to have this merged

**Comment by bithavoc** [comment]
Bumping for justice. Pretty please, merge!!!

**Comment by PibeG** [comment]
This is an important one.

**Comment by marcusellison** [comment]
Bumping for justice also.

**Comment by bithavoc** [comment]
@ianstormtaylor Hi, I'd like to buy you coffee, Paypal? Github Sponsors? 25$ Amazon Card?

If in the process we could also fix this issue, that'd be cool too.

**Review by ianstormtaylor — APPROVED** [review]

**Comment by marcusellison** [comment]
> @ianstormtaylor Hi, I'd like to buy you coffee, Paypal? Github Sponsors? $25 Amazon Card?
> 
> If in the process we could also fix this issue, that'd be cool too.

How do we get this merged after approval? :-)

**Comment by bithavoc** [comment]
Summoning @thesunny ☝️

**Comment by thesunny** [comment]
@marcusellison @bithavoc 

So, the good news is that we have a few new maintainers with publish ability in addition to Ian and Cameron now.

This includes @BrentFarese and @timbuckley. I've suggested to @BrentFarese when he asked about merging, to not merge until we have the new TypeScript types used against some real apps because it's a big PR.

I would say the best way to get things moving is to try `slate@next` in your project and see if the types are okay. Then let us know. I suggested that if we get two instances of people saying it works, that we do a normal release and start adding merging PRs.

Things should move faster after the new TypeScript types are in.

**Comment by lukesmurray** [comment]
@thesunny are you comfortable with typescript types as they currently stand?

**Comment by rockettomatooo** [comment]
Just noting, that there are similar PRs: #4137 and #4080.

**Comment by ianstormtaylor** [comment]
Thank you @ryanmitts!

**Pull request merged**
